components/web_discovery/browser/web_discovery_service.cc

```diff
@@ -0,0 +1,117 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/web_discovery/browser/web_discovery_service.h"
+
+#include <utility>
+
+#include "base/feature_list.h"
+#include "base/functional/bind.h"
+#include "brave/components/constants/pref_names.h"
+#include "brave/components/web_discovery/browser/pref_names.h"
+#include "brave/components/web_discovery/browser/server_config_loader.h"
+#include "brave/components/web_discovery/common/features.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+#include "extensions/buildflags/buildflags.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/service_manager/public/cpp/interface_provider.h"
+
+namespace web_discovery {
+
+WebDiscoveryService::WebDiscoveryService(
+    PrefService* local_state,
+    PrefService* profile_prefs,
+    base::FilePath user_data_dir,
+    scoped_refptr<network::SharedURLLoaderFactory> shared_url_loader_factory)
+    : local_state_(local_state),
+      profile_prefs_(profile_prefs),
+      user_data_dir_(user_data_dir),
+      shared_url_loader_factory_(shared_url_loader_factory) {
+#if BUILDFLAG(ENABLE_EXTENSIONS)
+  if (profile_prefs_->GetBoolean(kWebDiscoveryExtensionEnabled)) {
+    profile_prefs_->ClearPref(kWebDiscoveryExtensionEnabled);
+    profile_prefs_->SetBoolean(kWebDiscoveryNativeEnabled, true);
+  }
+#endif
+
+  pref_change_registrar_.Init(profile_prefs);
+  pref_change_registrar_.Add(
+      kWebDiscoveryNativeEnabled,
+      base::BindRepeating(&WebDiscoveryService::OnEnabledChange,
+                          base::Unretained(this)));
+
+  if (profile_prefs_->GetBoolean(kWebDiscoveryNativeEnabled)) {
+    Start();
+  }
+}
+
+WebDiscoveryService::~WebDiscoveryService() = default;
+
+void WebDiscoveryService::RegisterLocalStatePrefs(
+    PrefRegistrySimple* registry) {
+  registry->RegisterTimePref(kPatternsRetrievalTime, {});
+}
+
+void WebDiscoveryService::RegisterProfilePrefs(PrefRegistrySimple* registry) {
+  registry->RegisterBooleanPref(kWebDiscoveryNativeEnabled, false);
+  registry->RegisterDictionaryPref(kAnonymousCredentialsDict);
+  registry->RegisterStringPref(kCredentialRSAPrivateKey, {});
+  registry->RegisterStringPref(kCredentialRSAPublicKey, {});
+}
+
+void WebDiscoveryService::SetExtensionPrefIfNativeDisabled(
+    PrefService* profile_prefs) {
+#if BUILDFLAG(ENABLE_EXTENSIONS)
+  if (!base::FeatureList::IsEnabled(features::kWebDiscoveryNative) &&
+      profile_prefs->GetBoolean(kWebDiscoveryNativeEnabled)) {
```

Comment:
this seems overly complex, why do we have a separate pref for native and extension? Shouldn't there just be a single pref to enable it and the feature controls whether it's native or extension?

Reply:
the extension uses the `settingsPrivate` API to check the pref setting. i created another pref to avoid enabling the original extension code and the native component at the same time. if there's a way for the extension to check the feature state so that it won't start when the feature is enabled, let me know

Reply:
I forgot this is part of the brave extension, that makes it a bit more complicated but we're already making some custom apis available to the brave extension so we can add something for web discovery. I wouldn't bother making something for feature checks in general since the rest of the brave extension is going away and then we can just selectively enable/disable web discovery as its own extension if it still exists. Also right now I think if you start with the web discovery extension enabled and then enabled this flag and disable it again, web discovery will be completely disabled

Reply:
created a custom API, ended up being easier than expected, thanks